### PR TITLE
Add RequestParser trait to allow retrieving swagger information from …

### PR DESCRIPTION
**Pull request by mthebridge**
…other middlewares.

As summary.  This trait can then be implemented by `swagger-codegen`, allowing middleware services (or client wrappers) to get the operation from a request before the generated code actually handles it, for logging or other tracking purposes.

### CODE REVIEW

File: src/request_parser.rs [hunk 1]
```diff
@@ -0,0 +1,16 @@
...
+/// This allows other middlewares to retrieve API-related information from a request that
+/// may not have been handled by the autogenerated API code yet.   For example, a statistics
+/// tracking service may wish to use this to count requests per-operation.
+///
+/// The trait is automatically implemented by swagger-codegen.
+pub trait RequestParser {
+    /// Retrieve the Swagger operation identifier that matches this request.
+    ///
+    /// Returns `Err(())` if this request does not match any known operation on this API.
+    fn get_operation(req: &Request) -> Result<&'static str, ()>;
```
Comment (mirw) [inline src/request_parser.rs]: Should this be `get_operation_id`?  This gives flexibility if we need to return a parsed operation struct in future.

Comment (mthebridge) [reply src/request_parser.rs]: Works for me.  In fact, is it more Rusty to drop the `get` and just call it `operation_id`?

**Comment by mthebridge** [comment]
@bjgill Bit of both - the plan is to enhance codegen, yes, but the codegen needs to implemen a trait that's defined once, and surely better to do that here?  I didn't think codegen had any non-autogenerated Rust code?

I'll update the Changlog, thanks.

**Comment by mthebridge** [comment]
Markups made, thanks both.

**Comment by bjgill** [comment]
Thanks. I'll merge this and publish alongside https://github.com/Metaswitch/swagger-rs/pull/34

**Comment by bjgill** [comment]
Thanks @mirw  and @mthebridge. I assume this is ready to go now. Please shout if not - otherwise I'll merge and publish once Travis gives it the green light.

**Comment by bjgill** [comment]
Published in 0.12.1

**Pull request merged**